src/subclass/output_stream.rs

```diff
@@ -0,0 +1,381 @@
+// Copyright 2019, The Gtk-rs Project Developers.
+// See the COPYRIGHT file at the top-level directory of this distribution.
+// Licensed under the MIT license, see the LICENSE file or <http://opensource.org/licenses/MIT>
+
+use gio_sys;
+use glib_sys;
+
+use glib::subclass::prelude::*;
+use glib::translate::*;
+
+use glib::Error;
+
+use Cancellable;
+use InputStream;
+use OutputStream;
+use OutputStreamClass;
+use OutputStreamSpliceFlags;
+
+use std::mem;
+use std::ptr;
+
+pub trait OutputStreamImpl: OutputStreamImplExt + Send + 'static {
+    fn write(
+        &self,
+        stream: &OutputStream,
+        buffer: &[u8],
+        cancellable: Option<&Cancellable>,
+    ) -> Result<usize, Error> {
+        self.parent_write(stream, buffer, cancellable)
+    }
+
+    fn close(&self, stream: &OutputStream, cancellable: Option<&Cancellable>) -> Result<(), Error> {
+        self.parent_close(stream, cancellable)
+    }
+
+    fn flush(&self, stream: &OutputStream, cancellable: Option<&Cancellable>) -> Result<(), Error> {
+        self.parent_flush(stream, cancellable)
+    }
+
+    fn splice(
+        &self,
+        stream: &OutputStream,
+        input_stream: &InputStream,
+        flags: OutputStreamSpliceFlags,
+        cancellable: Option<&Cancellable>,
+    ) -> Result<usize, Error> {
+        self.parent_splice(stream, input_stream, flags, cancellable)
+    }
+}
+
+pub trait OutputStreamImplExt {
+    fn parent_write(
+        &self,
+        stream: &OutputStream,
+        buffer: &[u8],
+        cancellable: Option<&Cancellable>,
+    ) -> Result<usize, Error>;
+
+    fn parent_close(
+        &self,
+        stream: &OutputStream,
+        cancellable: Option<&Cancellable>,
+    ) -> Result<(), Error>;
+
+    fn parent_flush(
+        &self,
+        stream: &OutputStream,
+        cancellable: Option<&Cancellable>,
+    ) -> Result<(), Error>;
+
+    fn parent_splice(
+        &self,
+        stream: &OutputStream,
+        input_stream: &InputStream,
+        flags: OutputStreamSpliceFlags,
+        cancellable: Option<&Cancellable>,
+    ) -> Result<usize, Error>;
+}
+
+impl<T: OutputStreamImpl + ObjectImpl> OutputStreamImplExt for T {
+    fn parent_write(
+        &self,
+        stream: &OutputStream,
+        buffer: &[u8],
+        cancellable: Option<&Cancellable>,
+    ) -> Result<usize, Error> {
+        unsafe {
+            let data = self.get_type_data();
+            let parent_class = data.as_ref().get_parent_class() as *mut gio_sys::GOutputStreamClass;
+            let f = (*parent_class)
+                .write_fn
+                .expect("No parent class implementation for \"write\"");
+            let mut err = ptr::null_mut();
+            let res = f(
+                stream.to_glib_none().0,
+                buffer.as_ptr() as *mut u8,
```

Comment:
Should buffer be a `&mut [u8]` then? We're modifying content of a constant slice.

Reply:
We're not modifying it. It's just like `write()` in the `Write` trait. The bytes are written out to somewhere else. It's missing the `const` in C, that's the only problem here.

Reply:
Maybe use `mut_override`